// laocheng/Template.js
const lineReader = require('line-reader');
const fs=require('fs');
var readOver=0;
var _template='';
var _compile='';
function template(infile,outfile,template,compile,callback){
    _template=template;
    _compile=compile;
    readLine(infile,outfile);
    function readLine(infile,outfile) {
        readOver++;
        var fileContent = '';
        var _node=false;
        lineReader.eachLine(infile, function (line, last) {
            //转义字符
            line = changeY(line);
            //替换include文件
            line = changeInclude(line);
            //替换if，else，elseif
            line = changeIf(line);
            //替换permit
            line = changePermit(line);
            //替换for
            line = changeFor(line);
            //替换变量
            line = change$(line);
            //替换node标签
            line=changeNode(line);
            if(typeof line=='boolean'){
                _node=line;
                if(line==true){
                    fileContent+='{'+ '\n';
                }else{
                    fileContent+='}'+ '\n';
                }
                return;
            }
            if(!_node){
                //翻译页面
                line = resWrite(line);
            }else{
                line = line.replace(/\\"/g,'"');
            }
            if(line.trim()!=''){
                fileContent += line + '\n';
            }
            if (last) {
                readOver--;
                fileContent='const Common=require("'+__dirname.replace(/\\/g,'/')+'/Common");\nfunction temp(resp,val){\nwith(val){\n'+fileContent+'}\n}\nmodule.exports=temp;';
                fs.writeFileSync(outfile, fileContent,'utf-8');
                if(_node){
                    console.log('<node>标签未结尾，将会导致致命错误，请检查！')
                }
                if(readOver==0){
                    callback();
                }
            }
        });
    }

    function change$(line){
        var reg=new RegExp(/\${.+?}/g);
        var result;
        while ((result=reg.exec(line))!= null){
             var m=result[0].slice(2, result[0].length-1);
             line=line.replace(result[0],'"+('+m+')+"');
        }
        return line;
    }

    function resWrite(line){
        line='resp.write("'+line+'\\n");';
        return line;
    }

    function changeY(line){
        line=line.replace(/\"/g,'\\"');
        return line;
    }

    function changeIf(line){
        {
            var reg = new RegExp(/<if[ ]+test[ ]*=[ ]*\\"[ ]*\${.*?}[ ]*\\"[ ]*>/g);
            var result;
            while ((result = reg.exec(line)) != null) {
                var res = new RegExp(/\${.*?}/);
                var rm = res.exec(result[0]);
                if (rm != null) {
                    line = line.replace(result[0], '");if(' + rm[0].slice(2, rm[0].length - 1) + '){resp.write("');
                }
            }
        }
        {
            var reg = new RegExp(/<elseif[ ]+test[ ]*=[ ]*\\"[ ]*\${.*?}[ ]*\\"[ ]*\/>/g);
            var result;
            while ((result = reg.exec(line)) != null) {
                var res = new RegExp(/\${.*?}/);
                var rm = res.exec(result[0]);
                if (rm != null) {
                    line = line.replace(result[0], '");}else if(' + rm[0].slice(2, rm[0].length - 1) + '){resp.write("');
                }
            }
        }
        line=line.replace(/<else\/>/g,'");}else{resp.write("');
        line=line.replace(/<\/if>/g,'");}resp.write("');
        return line;
    }

    function changeFor(line){
        {
            var reg = new RegExp(/<for[ ]+from[ ]*=[ ]*\\"[ ]*\${.*?}[ ]*\\".*?>/g);
            var result;
            while ((result = reg.exec(line)) != null) {
                var res = new RegExp(/\${.*?}/);
                var rm = res.exec(result[0]);
                var r=rm[0].slice(2, rm[0].length - 1);
                var val=new RegExp(/value[ ]*=[ ]*\\"[ ]*[a-z0-9A-Z_]+?[ ]*\\"/);
                var v = val.exec(result[0]);
                var p=v[0].slice(v[0].indexOf('\"')+1, v[0].lastIndexOf('\"')-1).trim();
                var key=new RegExp(/key[ ]*=[ ]*\\"[ ]*[a-z0-9A-Z_]+?[ ]*\\"/);
                var k = key.exec(result[0]);
                var q=k[0].slice(k[0].indexOf('\"')+1, k[0].lastIndexOf('\"')-1).trim();
                if (rm != null) {
                    line = line.replace(result[0], '");for(var key in '+r+'){val.put("'+p+'",'+r+'[key]);val.put("'+q+'",key);resp.write("');
                }

            }
        }
        line=line.replace(/<\/for>/g,'");}resp.write("');
        return line;
    }

    function changeInclude(line){
        var reg = new RegExp(/<include[ ]*file[ ]*=[ ]*\\"[ ]*.*?[ ]*\\"[ ]*\/>/g);
        var result;
        while ((result = reg.exec(line)) != null) {
            var res = new RegExp(/file[ ]*=[ ]*\\"[ ]*[a-z0-9A-Z_\.]+?[ ]*\\"/);
            var rm = res.exec(result[0]);
            if (rm != null) {
                var q=rm[0].slice(rm[0].indexOf('\"')+1, rm[0].lastIndexOf('\"')-1).trim();
                var ql=q.slice(0,q.lastIndexOf('.'));
                readLine(_template+"/"+q,_compile+"/"+ql+".js");
                line=line.replace(result[0],'");require("./'+ql+'")(resp,val);resp.write("');
            }
        }
        return line;
    }

    function changeNode(line){
        if(line.indexOf('<node>')!=-1){
            return true;
        }else if(line.indexOf('<\/node>')!=-1){
            return false;
        }else{
            return line;
        }
    }

    function changePermit(line){
        var reg=new RegExp(/<permit[ ]*id[ ]*=[ ]*\\"[ ]*\d+[ ]*\\"[ ]*>/g);
        var result;
        while ((result = reg.exec(line)) != null) {
            var res = new RegExp(/id[ ]*=[ ]*\\"[ ]*\d+[ ]*\\"/);
            var rm = res.exec(result[0]);
            if (rm != null) {
                var p=rm[0].slice(rm[0].indexOf('\"')+1, rm[0].lastIndexOf('\"')-1).trim();
                line = line.replace(result[0], '");if(Common.permit('+p+',PERMIT)){resp.write("');
            }
        }
        line=line.replace(/<\/permit>/g,'");}resp.write("');
        return line;
    }
}
module.exports=template;
